### PR TITLE
6950 ARC should cache compressed data

### PR DESCRIPTION
**Pull request by grwilson**
Reviewed by: Prakash Surya prakash.surya@delphix.com
Reviewed by: Dan Kimmel dan.kimmel@delphix.com
Reviewed by: Matt Ahrens mahrens@delphix.com
Reviewed by: Paul Dagnelie pcd@delphix.com

This review covers the reading and writing of compressed arc headers, sharing data between the arc_hdr_t and the arc_buf_t, and the implementation of a new dbuf cache to keep frequently access data uncompressed. 

I've added a new member to l1 arc hdr called b_pdata. The b_pdata always hangs off the arc_buf_hdr_t (if an L1 hdr is in use) and points to the physical block for that DVA. The physical block may or may not be compressed. If compressed arc is enabled and the block on-disk is compressed, then the b_pdata will match the block on-disk and remain compressed in memory. If the block on disk is not compressed, then neither will the b_pdata. Lastly, if compressed arc is disabled, then b_pdata will always be an uncompressed version of the on-disk block.

Typically the arc will cache only the arc_buf_hdr_t and will aggressively evict any arc_buf_t's that are no longer referenced. This means that the arc will primarily have compressed blocks as the arc_buf_t's are considered overhead and are always uncompressed. When a consumer reads a block we first look to see if the arc_buf_hdr_t is cached. If the hdr is cached then we allocate a new arc_buf_t and decompress the b_pdata contents into the arc_buf_t's b_data. If the hdr already has a arc_buf_t, then we will allocate an additional arc_buf_t and bcopy the uncompressed contents from the first arc_buf_t to the new one.

Writing to the compressed arc requires that we first discard the b_pdata since the physical block is about to be rewritten. The new data contents will be passed in via an arc_buf_t (uncompressed) and during the I/O pipeline stages we will copy the physical block contents to a newly allocated b_pdata.

When an l2arc is inuse it will also take advantage of the b_pdata. Now the l2arc will always write the contents of b_pdata to the l2arc. This means that when compressed arc is enabled that the l2arc blocks are identical to those stored in the main data pool. This provides a significant advantage since we can leverage the bp's checksum when reading from the l2arc to determine if the contents are valid. If the compressed arc is disabled, then we must first transform the read block to look like the physical block in the main data pool before comparing the checksum and determining it's valid.

Upstreams bug:
DLPX-34721 track arc buf compressed size
DLPX-37223 ARC should cache compressed data
DLPX-37326 arc_state_t should use refcounts to track evictable space
DLPX-37633 disabling compressed arc with l2arc can cause a panic
DLPX-37644 dmu should cache frequently used dbufs
DLPX-37645 arc should do more data sharing
DLPX-38183 Kernel heap corruption detected in kmem_slab_free
DLPX-38295 add verification of allocations and frees for arc data buffers
DLPX-38547 delphix engine hang
DLPX-38547 delphix engine hang (fix static global)
DLPX-38882 system hung waiting on free segment
DLPX-39741 arc_cksum_equal fails to handle gang blocks
DLPX-39793 arc_stats have unused kstats
DLPX-39794 Stack overflow out of dbuf_rele
DLPX-39860 want a way to ensure that b_flags are updated in a thread safe manner
DLPX-40013 memory leak in dbuf_destroy()
DLPX-40457 cherry pick upstream: 6220 memleak in l2arc on debug build
DLPX-43482 prefetched dbufs are not being cached correctly
DLPX-44361 l2arc_write_buffers can use stale data
QA-4903 Performance regression suite for zfstest
QA-4904 testrunner should be able to specify tests that can't timeout

### CODE REVIEW

File: usr/src/test/test-runner/cmd/run.py [hunk 1]
```diff
@@ -144,13 +145,16 @@ class Cmd(object):
     def __init__(self, pathname, outputdir=None, timeout=None, user=None):
         self.pathname = pathname
         self.outputdir = outputdir or 'BASEDIR'
-        self.timeout = timeout or 60
+        self.timeout = timeout
         self.user = user or ''
         self.killed = False
         self.result = Result()
 
+        if self.timeout == None:
```
Comment (richardelling) [inline usr/src/test/test-runner/cmd/run.py]: this is better as:
if not self.timeout:

File: usr/src/test/test-runner/cmd/run.py [hunk 2]
```diff
@@ -144,13 +145,16 @@ class Cmd(object):
...
+            self.timeout = 60
+
```
Comment (richardelling) [inline usr/src/test/test-runner/cmd/run.py]: if self.timeout is not an int, then you'll get ValueError thrown later at line 236 or 326
it is easy to check here first to ensure it is always convertible to int, something like:
```
try:
    self.timeout = int(self.timeout)
except ValueError as exc:
    # complain
    self.timeout = 60
```
then you'll be able to remove the `int(self.timeout)` later

File: usr/src/test/test-runner/cmd/run.py [hunk 3]
```diff
@@ -144,13 +145,16 @@ class Cmd(object):
     def __init__(self, pathname, outputdir=None, timeout=None, user=None):
         self.pathname = pathname
         self.outputdir = outputdir or 'BASEDIR'
```
Comment (richardelling) [inline usr/src/test/test-runner/cmd/run.py]: nit...
I'm not sure why the original author didn't set these defaults in the `def __init__` above, it is more friendly to do so and useful when using an IDE. Doing so avoids the need for the `or blah` here.

File: usr/src/test/zfs-tests/runfiles/perf-regression.run [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+pre_user = root
+user = root
+timeout = 0
+post_user = root
+post = cleanup
+outputdir = /var/tmp/test_results
+
+[/opt/zfs-tests/tests/perf/regression]
+tests = ['sequential_writes', 'sequential_reads', 'sequential_reads_cached',
+    'sequential_reads_cached_clone, 'random_reads', 'random_writes',
```
Comment (don-brady) [inline usr/src/test/zfs-tests/runfiles/perf-regression.run]: `sequential_reads_cached_clone` appears to be missing ending quote mark

Comment (grwilson) [inline usr/src/test/zfs-tests/runfiles/perf-regression.run]: @don-brady I missed one of the commits and have fixed it. Thanks for pointing this out.

File: usr/src/test/test-runner/cmd/run.py (hunk 3 above)
Comment (jwk404) [inline usr/src/test/test-runner/cmd/run.py]: I'm not sure either. :-) Seems like the right thing to do.

File: usr/src/test/test-runner/cmd/run.py (hunk 1 above)
Comment (jwk404) [inline usr/src/test/test-runner/cmd/run.py]: Normally, I'd agree, but if we use `if not self.timeout` then when timeout is 0 (meaning infinite) we'll reset it to 60, which we don't want.

File: usr/src/test/test-runner/cmd/run.py (hunk 2 above)
Comment (jwk404) [inline usr/src/test/test-runner/cmd/run.py]: I don't think checking for TypeError here is required. At this point, timeout can only be either None or an int, and if it's the former we change it to an int.

**Comment by don-brady** [comment]
We have the perf regression tests and the compressed arc code from the PR running now on zfsonlinux.  So a good sign.  Hope to finish up formal review soon.

**Comment by don-brady** [comment]
@grwilson  The performance regression tests **LGTM**.

A few minor nits on the ARC changes below, otherwise those are also **LGTM**

Comments for `arc.c`: 

```
@@ -225,9 +351,11 @@ uint64_t zfs_arc_meta_min = 0;
 int zfs_arc_grow_retry = 0;
 int zfs_arc_shrink_shift = 0;
 int zfs_arc_p_min_shift = 0;
-int zfs_disable_dup_eviction = 0;
 int zfs_arc_average_blocksize = 8 * 1024; /* 8KB */
 
+boolean_t zfs_dup_eviction_enabled = B_TRUE;
+boolean_t zfs_compressed_arc_enabled = B_TRUE;
+
 /*
  * Note that buffers can be in one of 6 states:
  *     ARC_anon        - anonymous (discussed below)
```
Note that `zfs_dup_eviction_enabled` appears to be an orphaned tunable. It was originally a conversion from `zfs_disable_dup_eviction` but regardless it was only used in `arc_buf_eviction_needed()` which is now gone.
 ______________________________________________
 
```
@@ -216,6 +337,11 @@ static int arc_dead;
 static boolean_t arc_warm;
 
 /*
+ * log2 fraction of the zio arena to keep free.
+ */
+int arc_zio_arena_free_shift = 2;
+
+/*
  * These tunables are for performance analysis.
  */
 uint64_t zfs_arc_max;
```
Is `arc_zio_arena_free_shift` intended to be an external tunable?  If so, ideally it would have
a `zfs_` prefix name like the other external zfs tunables.
 ______________________________________________
 
```
@@ -757,20 +896,37 @@ struct arc_buf_hdr {
 <snip>
+        *
+        * This isn't a problem in practice, since the maximum size of a
+        * buffer is limited to 128K, so we never need to store 32MB in
+        * this field. Even in the upstream illumos code base, the
+        * maximum size of a buffer is limited to 16MB.
 <snip>
```
Now that OpenZFS supports 16MB, is the 128K comment stale?

**Comment by don-brady** [comment]
One more issue found when porting to Linux.  The dbuf kmem cache constructor, `dbuf_cons()`, is missing a call to `multilist_link_init()`.  This was leading to false positive `multilist_link_active()` results.

**Comment by grwilson** [comment]
@don-brady Thanks for looking at this pull request. I'm addressing all but the arc_zio_free_shift since it's not designed to be used externally and already exists in openzfs (introduced by  7016 arc_available_memory is not 32-bit safe).

**Comment by benrubson** [comment]
/!\ It seems that this PR leaves the L2ARC uncompressed when the pool (and so the ARC) has compression off.
Up to now, even when the pool has compression off, L2ARC is always compressed, which is nice because it can then contain more data.
We can have compression off because data compression ratio will not be so good, but it can be interesting to have L2ARC compressed if for example it only contains metadata.
This is my use case, pool compression off (because my users' data has a really bad compression ratio), secondarycache=metadata, L2ARC is 130GB, only 46GB on SSD disk.

Could we then think about re-enabling L2ARC compression please ?

Thank you very much 👍 

Ben

**Comment by benrubson** [comment]
And here is a comment for arc.c :

```
if ((write_asize + HDR_GET_LSIZE(hdr)) > target_sz) {
	full = B_TRUE;
	mutex_exit(hash_lock);
	ARCSTAT_BUMP(arcstat_l2_write_full);
	break;
}
```

I think the following is better ?
`if ((write_asize + arc_hdr_size(hdr)) > target_sz) {`

Thank you !

Ben

**Comment by grwilson** [comment]
@benrubson: metadata is always compressed (unless you have overridden this) so it will be compressed on L2ARC and ARC.

**Comment by benrubson** [comment]
George yes, my mistake, sorry for this, perfect that metadata is always compressed.
Thank you 👍

**Comment by benrubson** [comment]
In addition to my code proposal above, in arc.c, I would also do the following modifications :

From :
```
ARCSTAT_INCR(arcstat_l2_write_buffer_bytes_scanned,
    HDR_GET_LSIZE(hdr));
(...)
passed_sz += HDR_GET_LSIZE(hdr);
if (passed_sz > headroom) {
```

To :
```
ARCSTAT_INCR(arcstat_l2_write_buffer_bytes_scanned,
    arc_hdr_size(hdr));
(...)
passed_sz += arc_hdr_size(hdr);
if (passed_sz > headroom) {
```

Don't we prefer to know (and to work on) how much memory has been scanned, not how much uncompressed data it represents ?
In addition, as the function returns the sum of arc_hdr_size, I think this is more consistent.

Many thanks again !

Ben

**Comment by scineram** [comment]
@grwilson If I understand correctly with compressed ARC turned off an L2ARC read will compress the block just to compare checksum with that of the blockpointer. Wouldn't it make sense to always have L2ARC compressed to save space and on reads have less latency with decompression faster than compression?

**Comment by grwilson** [comment]
@scineram If the compressed ARC is disabled (different thank disabling compression on the pool or dataset) and you have compression enabled on a dataset, then yes, the reading of the block from the l2arc device will require the block to be compressed so that we can compare that to the blkptr's checksum. We do not expect this to be a commonly used configuration. Do you have a use case where you're disabling compressed ARC but still leaving compression enabled in the pool/dataset?

**Pull request closed without merging**